threedi_models_and_simulations/widgets/substance_concentrations.py

```diff
@@ -0,0 +1,109 @@
+        font = QFont("Segoe UI", 10, QFont.Normal)
+        for i, substance in enumerate(self.substances):
+            name = substance["name"]
+            label = QLabel(f"{name}:")
+            label.setMinimumWidth(100)
+            label.setFont(font)
+            line_edit = QLineEdit()
+            line_edit.setObjectName("le_substance_" + name)
+            line_edit.setReadOnly(True)
+            line_edit.setFrame(False)
+            line_edit.setFont(font)
+            line_edit.setStyleSheet("background-color: white")
+            upload_button = QPushButton("Upload CSV")
+            upload_button.setObjectName("pb_substance_" + name)
+            upload_button.setMinimumWidth(100)
+            upload_button.setFont(font)
+            horizontal_layout = QHBoxLayout()
+            horizontal_layout_widget = QWidget()
+            horizontal_layout_widget.setLayout(horizontal_layout)
+            horizontal_layout.setContentsMargins(0, 0, 9, 0)
+            horizontal_layout.addWidget(label)
+            horizontal_layout.addWidget(line_edit)
+            horizontal_layout.addWidget(upload_button)
+            layout.addWidget(horizontal_layout_widget, i, 0)
+        self.connect_substance_upload_signals()
+
+    def connect_substance_upload_signals(self):
+        """Connect substance upload signals."""
+        for substance in self.substances:
+            name = substance["name"]
+            upload_button = self.groupbox.findChild(QPushButton, "pb_substance_" + name)
+            upload_button.clicked.connect(partial(self.load_substance_csv, name))
+
+    def load_substance_csv(self, name):
+        """Load substance CSV file."""
+        substances, filename = self.open_substance_upload_dialog(name)
+        if not filename:
+            return
+        le_substance = self.groupbox.findChild(QLineEdit, "le_substance_" + name)
+        le_substance.setText(filename)
+        self.substance_concentrations.update(substances)
+
+    def open_substance_upload_dialog(self, name):
+        """Open dialog for selecting CSV file with laterals."""
+        last_folder = QSettings().value("threedi/last_substances_folder", os.path.expanduser("~"), type=str)
+        file_filter = "CSV (*.csv );;All Files (*)"
+        filename, __ = QFileDialog.getOpenFileName(
+            self, f"Substance Concentrations for {name}", last_folder, file_filter
+        )
+        if len(filename) == 0:
+            return None, None
+        QSettings().setValue("threedi/last_substances_folder", os.path.dirname(filename))
+        substances = {}
+        substance_list = []
+        with open(filename, encoding="utf-8-sig") as substance_file:
+            substance_reader = csv.reader(substance_file)
+            substance_list += list(substance_reader)
+        error_msg = self.handle_substance_errors(substance_list)
+        if error_msg is not None:
+            return None, None
+        for id, timeseries in substance_list:
```

Comment:
It's good practice to avoid shadowing built-in functions. Please change the name of the `id` variable.